pkg/ingester/ingester_test.go

```diff
@@ -5677,3 +5677,123 @@ func TestGetIgnoreSeriesLimitForMetricNamesMap(t *testing.T) {
 	cfg.IgnoreSeriesLimitForMetricNames = "foo, bar, ,"
 	require.Equal(t, map[string]struct{}{"foo": {}, "bar": {}}, cfg.getIgnoreSeriesLimitForMetricNamesMap())
 }
+
+func Test_Ingester_QueryOutOfOrder(t *testing.T) {
+	tests := map[string]struct {
+		queryFrom      int64
+		queryTo        int64
+		oooFirstSample int64
+		oooLastSample  int64
+		expPushError   bool
+	}{
+		"should return in order and out of order data": {
+			queryFrom:      math.MinInt64,
+			queryTo:        math.MaxInt64,
+			oooFirstSample: 70 * time.Minute.Milliseconds(),
+			oooLastSample:  99 * time.Minute.Milliseconds(),
+			expPushError:   false,
+		},
+		"if ooo samples go back past allowance writing should return an error": {
+			queryFrom:      math.MinInt64,
+			queryTo:        math.MaxInt64,
+			oooFirstSample: 10 * time.Minute.Milliseconds(), // Allowance is 30 min, first sample is at minute 100 so first ooo sample is too old
+			oooLastSample:  99 * time.Minute.Milliseconds(),
+			expPushError:   true,
+		},
+	}
+
+	// Configure ingester with OOO Allowance
+	cfg := defaultIngesterTestConfig(t)
+	cfg.BlocksStorageConfig.TSDB.OOOAllowance = 30 * time.Minute
+	cfg.BlocksStorageConfig.TSDB.OOOCapMin = 4
+	cfg.BlocksStorageConfig.TSDB.OOOCapMax = 32
+
+	// Create ingester
+	i, err := prepareIngesterWithBlocksStorage(t, cfg, nil)
```

Comment:
Just a note that, if you wanted to test more cases that does not result an error, we will need this ingester creation per test case so start with a clean TSDB. Not a problem right now since we have only 2 cases which do not depend on each other.